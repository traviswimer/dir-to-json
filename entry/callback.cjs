// This allows require() to work as expected when used in CommonJS
module.exports = require("../dist/callback.cjs").default;
