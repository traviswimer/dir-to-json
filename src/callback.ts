import dirToJson from "./main";
import { Options } from "./createDirectoryObject";

export const INVALID_CALLBACK_ERROR = `The "dir-to-json/callback" module requires a valid callback function`;

function callback(path: string, callbackFn: Function | undefined);
function callback(
	path: string,
	options: Options,
	callbackFn: Function | undefined
);
function callback(
	path: string,
	optionsOrCallbackFn: Function | Options | undefined,
	callbackFn?: Function
) {
	let options;
	if (typeof optionsOrCallbackFn === "function") {
		options = undefined;
		callbackFn = optionsOrCallbackFn;
	} else {
		options = optionsOrCallbackFn;
	}

	if (typeof callbackFn !== "function") {
		throw new Error(INVALID_CALLBACK_ERROR);
	}

	dirToJson(path, options)
		.then((dirTree) => {
			callbackFn(undefined, dirTree);
		})
		.catch((err) => {
			callbackFn(err);
		});
}

export default callback;
