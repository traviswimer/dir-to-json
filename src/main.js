var fs = require("fs");
var Q = require("q");

var createDirectoryObject = require("./createDirectoryObject");

module.exports = function (path, options, callbackParam) {
	var deferred = Q.defer();
	callbackParam = callbackParam || function () {};

	if (typeof options === "function") {
		callbackParam = options;
	}

	// Make promises and callbacks both available
	var callback = function (err, data) {
		if (err) {
			deferred.reject(err);
		} else {
			deferred.resolve(data);
		}

		callbackParam(err, data);
	};

	if (typeof path !== "string") {
		callback(new Error("Path parameter must be a string"));
		return deferred.promise;
	}

	createDirectoryObject(path, "", options)
		.then(function (obj) {
			callback(null, obj);
		})
		.catch(function (err) {
			callback(err);
		});

	return deferred.promise;
};
