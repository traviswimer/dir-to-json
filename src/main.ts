import createDirectoryObject, { Options } from "./createDirectoryObject";

export const INVALID_PATH_ERROR = `"path" parameter must be a string`;
export const INVALID_OPTIONS_ERROR = `"options" parameter must be an object`;
export default async function dirToJson(path: string, options: Options = {}) {
	if (typeof path !== "string") {
		return Promise.reject(INVALID_PATH_ERROR);
	}
	if (typeof options !== "object") {
		return Promise.reject(INVALID_OPTIONS_ERROR);
	}

	return await createDirectoryObject(path, "", options);
}
