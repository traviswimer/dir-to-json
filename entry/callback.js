import callback from "../dist/callback.js";
export default callback;
