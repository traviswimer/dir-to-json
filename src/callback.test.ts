import path from "path";

import callback, { INVALID_CALLBACK_ERROR } from "./callback";
import { INVALID_PATH_ERROR } from "./main";

test(`rejects if "callback" is not a valid function`, () => {
	// @ts-ignore: This is testing an invalid parameter
	return expect(() => callback("a/fake/path", null)).toThrow(
		INVALID_CALLBACK_ERROR
	);
});

test(`successfully generates directory tree with all valid parameters`, (done) => {
	const filePath = path.resolve(__dirname, "../fixtures/dir-5");
	callback(filePath, undefined, (err: any, data: any) => {
		expect(err).toBeUndefined();
		expect(data).toMatchSnapshot();
		done();
	});
});

test(`successfully generates directory tree with no "options" parameter provided`, (done) => {
	const filePath = path.resolve(__dirname, "../fixtures/dir-5");
	callback(filePath, (err: any, data: any) => {
		expect(err).toBeUndefined();
		expect(data).toMatchSnapshot();
		done();
	});
});

test(`callback receives errors when they occur`, (done) => {
	const filePath = path.resolve(__dirname, "../fixtures/dir-5");
	callback(
		// @ts-ignore: This is testing an invalid parameter
		{
			this: "is not a valid path string",
		},
		(err: any, data: any) => {
			expect(data).toBeUndefined();
			expect(err).toEqual(INVALID_PATH_ERROR);
			done();
		}
	);
});
