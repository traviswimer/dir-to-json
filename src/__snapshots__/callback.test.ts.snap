// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successfully generates directory tree with all valid parameters 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "name": "1-file.txt",
              "parent": "2-dir/1-dir",
              "path": "2-dir/1-dir/1-file.txt",
              "type": "file",
            },
            Object {
              "name": "2-file.txt",
              "parent": "2-dir/1-dir",
              "path": "2-dir/1-dir/2-file.txt",
              "type": "file",
            },
          ],
          "name": "1-dir",
          "parent": "2-dir",
          "path": "2-dir/1-dir",
          "type": "directory",
        },
        Object {
          "children": Array [
            Object {
              "name": "1-file.txt",
              "parent": "2-dir/3-dir",
              "path": "2-dir/3-dir/1-file.txt",
              "type": "file",
            },
            Object {
              "name": "2-file.txt",
              "parent": "2-dir/3-dir",
              "path": "2-dir/3-dir/2-file.txt",
              "type": "file",
            },
          ],
          "name": "3-dir",
          "parent": "2-dir",
          "path": "2-dir/3-dir",
          "type": "directory",
        },
        Object {
          "name": "2-file.txt",
          "parent": "2-dir",
          "path": "2-dir/2-file.txt",
          "type": "file",
        },
      ],
      "name": "2-dir",
      "parent": "",
      "path": "2-dir",
      "type": "directory",
    },
    Object {
      "name": "1-file.txt",
      "parent": "",
      "path": "1-file.txt",
      "type": "file",
    },
    Object {
      "name": "3-file.txt",
      "parent": "",
      "path": "3-file.txt",
      "type": "file",
    },
  ],
  "name": "dir-5",
  "parent": "..",
  "path": "",
  "type": "directory",
}
`;

exports[`successfully generates directory tree with no "options" parameter provided 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "name": "1-file.txt",
              "parent": "2-dir/1-dir",
              "path": "2-dir/1-dir/1-file.txt",
              "type": "file",
            },
            Object {
              "name": "2-file.txt",
              "parent": "2-dir/1-dir",
              "path": "2-dir/1-dir/2-file.txt",
              "type": "file",
            },
          ],
          "name": "1-dir",
          "parent": "2-dir",
          "path": "2-dir/1-dir",
          "type": "directory",
        },
        Object {
          "children": Array [
            Object {
              "name": "1-file.txt",
              "parent": "2-dir/3-dir",
              "path": "2-dir/3-dir/1-file.txt",
              "type": "file",
            },
            Object {
              "name": "2-file.txt",
              "parent": "2-dir/3-dir",
              "path": "2-dir/3-dir/2-file.txt",
              "type": "file",
            },
          ],
          "name": "3-dir",
          "parent": "2-dir",
          "path": "2-dir/3-dir",
          "type": "directory",
        },
        Object {
          "name": "2-file.txt",
          "parent": "2-dir",
          "path": "2-dir/2-file.txt",
          "type": "file",
        },
      ],
      "name": "2-dir",
      "parent": "",
      "path": "2-dir",
      "type": "directory",
    },
    Object {
      "name": "1-file.txt",
      "parent": "",
      "path": "1-file.txt",
      "type": "file",
    },
    Object {
      "name": "3-file.txt",
      "parent": "",
      "path": "3-file.txt",
      "type": "file",
    },
  ],
  "name": "dir-5",
  "parent": "..",
  "path": "",
  "type": "directory",
}
`;
